Userland/Libraries/LibTest/TestRunner.h

```diff
@@ -296,4 +320,118 @@ inline void TestRunner::print_test_results_as_json() const
     outln("{}", root.to_string());
 }
 
+inline void TestRunner::print_test_results_as_junit() const
+{
+    auto* junit_out = fopen(String::formatted("{}/{}", Core::StandardPaths::home_directory(), "junit.xml").characters(), "+w");
+
+    constexpr auto iso8601_format = "%Y-%m-dT%H:%M:%S%z"sv;
+
+    String hostname = "localhost"sv;
+    auto hostname_or_error = Core::System::gethostname();
+    if (!hostname_or_error.is_error())
+        hostname = hostname_or_error.release_value();
+    String timestamp = Core::DateTime::now().to_string(iso8601_format);
+
+    outln(junit_out, R"~~~(<?xml version="1.0" encoding="UTF-8"?>
+    <testsuites>)~~~");
+
+    auto const& suites = this->suites();
+
+    for (auto const& suite : suites) {
+        size_t num_failures = 0;
+        size_t num_skipped = 0;
+        size_t num_crashed = 0;
+
+        for (auto const& test : suite.tests) {
+            if (test.result == Test::Result::Skip)
+                ++num_skipped;
+            else if (test.result == Test::Result::Fail) {
+                ++num_failures;
+            } else if (test.result == Test::Result::Crashed) {
+                ++num_crashed;
+            }
+        }
+        // Just one "test" in this case
+        if (suite.tests.is_empty()) {
+            if (suite.most_severe_test_result == Test::Result::Fail)
+                ++num_failures;
+            else if (suite.most_severe_test_result == Test::Result::Skip)
+                ++num_skipped;
+            if (suite.most_severe_test_result == Test::Result::Crashed)
+                ++num_crashed;
+        }
+        String name = LexicalPath(suite.name).basename();
+        if (name == "__$$TOP_LEVEL$$__"sv)
+            name = String::empty();
+
+        auto path = LexicalPath::relative_path(suite.path, m_test_root);
+
+        outln(junit_out, R"~~~(<testsuite name="{}"
+    tests="{}"
+    failures="{}"
+    disabled="0"
+    skipped="{}"
+    errors="{}"
+    hostname="{}"
+    time="{}"
+    timestamp="{}"
+    >)~~~",
+            path,
+            suite.tests.size() ?: 1,
+            num_failures,
+            num_skipped,
+            num_crashed,
+            hostname,
+            suite.total_duration_us / 1'000'000,
+            timestamp);
+
+        auto result_to_status = [](Test::Result res) {
+            switch (res) {
+            case Test::Result::Pass:
+                return "run"sv;
+            case Test::Result::Skip:
+                return "disabled"sv;
+            case Test::Result::Crashed:
+                return "error"sv;
+            case Test::Result::Fail:
+                return "fail"sv;
+            }
+            VERIFY_NOT_REACHED();
+        };
+
+        for (auto const& test : suite.tests) {
+            StringView status = result_to_status(test.result);
+            outln(junit_out, R"~~~(<testcase name="{}" classname="{}" time="{}" status="{}">)~~~", test.name, name, double(test.duration_us) / 1'000'000.0, status);
```

Comment:
Function-style casts for when C-style casts are bad and `static_cast` is too ugly?